tests/common_tests/transport/redis_gateway/test_threading.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import absolute_import, unicode_literals
+
+import mock
+import threading
+import time
+import unittest
+
+from pysoa.common.metrics import NoOpMetricsRecorder
+from pysoa.common.transport.exceptions import MessageReceiveTimeout
+from pysoa.common.transport.redis_gateway.client import RedisClientTransport
+from pysoa.common.transport.redis_gateway.constants import REDIS_BACKEND_TYPE_STANDARD
+from pysoa.common.transport.redis_gateway.server import RedisServerTransport
+
+
+class _FakeBackend(object):
+    """
+    This fake Redis backend is not thread-safe. That's okay. We don't need it to be for this test, because we are
+    intentionally controlling the timing of things.
+    """
+    def __init__(self):
+        self._container = {}
+
+    def get_connection(self, *_):
+        return self
+
+    def send_message_to_queue(self, queue_key, message, *_, **__):
+        # print('\n    - send: {} / {} / {}\n'.format(queue_key, type(message), message))
+        self._container.setdefault(queue_key, list()).append(message)
+
+    def blpop(self, keys, *_, **__):
+        if self._container.get(keys[0]):
+            message = self._container[keys[0]].pop(0)
+            # print('\n    - receive: {} / {} / {}\t'.format(keys[0], type(message), message))
+            return [keys[0], message]
+        return None
+
+
+class _FakeEchoingServer(threading.Thread):
+    """
+    This is a super simple server that simply receives requests and immediately replies with responses.
+    """
+    def __init__(self, server_transport):
+        self._transport = server_transport
+        self._continue = True
+        self.error = None
+
+        super(_FakeEchoingServer, self).__init__()
+
+    def shutdown(self):
+        self._continue = False
+
+    def run(self):
+        while self._continue is True:
+            try:
+                # print('  - server receive')
+                request_id, meta, body = self._transport.receive_request_message()
+                # print('  - server send')
+                self._transport.send_response_message(request_id, meta, body)
+            except MessageReceiveTimeout:
+                time.sleep(0.05)
```

Comment:
I'm not sure if `freezegun` works fine with threads, but if it does, I'd much rather see `freezegun.tick()` instead of `time.sleep()` in a test. 

OTOH, consider this a non-blocking nitpick.

Reply:
Ehhh ... that seems like more trouble than it's worth. :-)